Add ShoppingCart model

dmarmul opened the pull request:
Added ShoppingCart and CartItem models, mappers, Dtos, service, repositories and controller for adding CartItem to ShoppingCart, update, delete and show items.

dmarmul committed:
Added ShoppingCart and CartItem models, mappers, Dtos, service, repositories and controller for adding CartItem to ShoppingCart, update, delete and show items.

ahoienko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,76 @@
+@RestController
+@RequestMapping("/cart")
+public class CartController {
+    private static final int BEARER_SUBSTRING = 7;
+
```
```suggestion

```

ahoienko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,76 @@
+    private static final int BEARER_SUBSTRING = 7;
+
+    private final CartService cartService;
+    private final JwtUtil jwtUtil;
+    private String userName;
+
+    @ModelAttribute
+    public void setUserName(@RequestHeader("Authorization") String token) {
+        this.userName = jwtUtil.getUsername(token.substring(BEARER_SUBSTRING));
+    }
```
```suggestion
```

ahoienko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,76 @@
+    @Operation(summary = "Add books to cart",
+            description = "Add books to the shopping cart."
+                    + " Fields bookId and quantity can't must be not less than 1, can't be null.")
+    public CartItemDto add(@RequestBody @Valid CreateCartRequestDto requestDto) {
```
```suggestion
    public CartItemDto add(@RequestBody @Valid CreateCartRequestDto requestDto, @AuthenticationPrincipal User user) {
```

ahoienko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,76 @@
+            description = "Add books to the shopping cart."
+                    + " Fields bookId and quantity can't must be not less than 1, can't be null.")
+    public CartItemDto add(@RequestBody @Valid CreateCartRequestDto requestDto) {
+        return cartService.add(requestDto, userName);
```
```suggestion
        return cartService.add(requestDto, user);
```

ahoienko commented on src/main/java/org/example/bookshop/dto/CartDto.java:
```diff
@@ -0,0 +1,13 @@
+
+@Getter
+@Setter
+public class CartDto {
```
```suggestion
public class ShoppingCartDto {
```

ahoienko commented on src/main/java/org/example/bookshop/mapper/ShoppingCartMapper.java:
```diff
@@ -0,0 +1,26 @@
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+
+@Mapper(config = MapperConfig.class)
```
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

ahoienko commented on src/main/java/org/example/bookshop/mapper/ShoppingCartMapper.java:
```diff
@@ -0,0 +1,26 @@
+public interface ShoppingCartMapper {
+    @Mapping(target = "cartItems", ignore = true)
+    CartDto toDto(ShoppingCart cart);
+
+    @AfterMapping
+    default void setCartDtoParam(@MappingTarget CartDto cartDto, ShoppingCart cart) {
+        CartItemMapper cartItemMapper = new CartItemMapperImpl();
+        cartDto.setUserId(cart.getUser().getId());
+        cartDto.setCartItems(cart.getCartItems().stream()
+                .map(cartItemMapper::toCartItemDto)
+                .collect(Collectors.toSet()));
+    }
```
```suggestion

```

ahoienko commented on src/main/java/org/example/bookshop/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @ManyToOne
+    private Book book;
+    @Column(nullable = false)
+    @Min(1)
```
```suggestion
```

ahoienko commented on src/main/java/org/example/bookshop/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id")
+    private ShoppingCart shoppingCart;
+    @ManyToOne
+    private Book book;
+    @Column(nullable = false)
```
Please make fetch type Lazy

ahoienko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,76 @@
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get shopping cart",
+            description = "Retrieve user's shopping cart")
+    public CartDto getAll() {
```
```suggestion
    public CartDto get() {
```

ahoienko commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+    private final UserRepository userRepository;
+
+    @Override
+    public CartDto findAll(String username) {
```
```suggestion
    public CartDto get(String username) {
```

ahoienko commented on src/main/java/org/example/bookshop/dto/CreateCartRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+@Setter
+public class CreateCartRequestDto {
```
```suggestion
public class CartItemRequestDto {
```

ahoienko reviewed (changes requested):
Good job! Please check comments

dmarmul committed:
Rename some classes and methods.
Replaced afterMapping in ShoppingCartMapper and CartItemMapper, added uses annotation.
Added AuthenticationPrincipal annotation in CartController endpoints.
Fixes some mistakes.

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,68 @@
+    @Operation(summary = "Add books to cart",
+            description = "Add books to the shopping cart."
+                    + " Fields bookId and quantity can't must be not less than 1, can't be null.")
+    public CartItemDto add(@RequestBody @Valid CartItemRequestDto requestDto,
```
```suggestion
    public ShoppingCartDto add(@RequestBody @Valid CartItemRequestDto requestDto,
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -0,0 +1,68 @@
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Update the books quantity",
+            description = "Update the books quantity in shopping cart")
+    public CartItemDto update(@RequestBody @Valid CartUpdateRequestDto requestDto,
```
```suggestion
    public ShoppingCartDto update(@RequestBody @Valid CartUpdateRequestDto requestDto,
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/dto/CartItemRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+    @NotNull
+    @Min(1)
+    private Long bookId;
+    @NotNull
```
int type can't be null
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/dto/CartItemRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+@Setter
+public class CartItemRequestDto {
+    @NotNull
+    @Min(1)
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/dto/CartItemRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+    @Min(1)
+    private Long bookId;
+    @NotNull
+    @Min(1)
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/dto/CartUpdateRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+@Setter
+public class CartUpdateRequestDto {
+    @NotNull
+    @Min(1)
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/CartItem.java:
```diff
@@ -0,0 +1,32 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id")
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/CartItem.java:
```diff
@@ -0,0 +1,32 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id")
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/CartItem.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@Getter
+@Setter
+@ToString
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/ShoppingCart.java:
```diff
@@ -0,0 +1,40 @@
+@Entity
+@Getter
+@Setter
+@ToString
```
ToString leads to exception in such entities
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/ShoppingCart.java:
```diff
@@ -0,0 +1,40 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
```
```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/ShoppingCart.java:
```diff
@@ -0,0 +1,40 @@
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/repository/CartRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface CartRepository extends JpaRepository<ShoppingCart, Long> {
+    @EntityGraph(attributePaths = {"user", "user.roles", "cartItems", "cartItems.book"})
+    @NonNull
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Override
+    public CartItemDto add(CartItemRequestDto requestDto, User user) {
```
```suggestion
    public ShoppingCartDto add(CartItemRequestDto requestDto, User user) {
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Override
+    public CartItemDto update(CartUpdateRequestDto requestDto, Long id, User user) {
```
```suggestion
    public ShoppingCartDto update(CartUpdateRequestDto requestDto, Long id, User user) {
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+    private ShoppingCart findShoppingCart(User user) {
+        return cartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                "Can't found cartRepository for user :" + user.getEmail())
```
```suggestion
                "Can't find cartRepository for user :" + user.getEmail())
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public void delete(Long id, User user) {
+        if (findCartItemById(findShoppingCart(user), id) != null) {
```
it's not readable
find shop cart `findByUserId`
find cartItem `findByIdAndShoppingCartId`

remove cartItem

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+@RequiredArgsConstructor
+@Service
+public class CartServiceImpl implements CartService {
```
Think where `@Transactional` is needed

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public CartItemDto add(CartItemRequestDto requestDto, User user) {
+        CartItem cartItem = cartItemMapper.toCartItem(requestDto);
```
it's better to find book, find shop cart, check if cart already include the same book, if yes - increment quantity, if no - add book to cart

It's better to save shop cart at the end of method

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -31,6 +34,11 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
-        return userMapper.toDto(userRepository.save(user));
+        user = userRepository.save(user);
+
+        ShoppingCart shoppingCart = new ShoppingCart();
```
create method `createShoppingCart(User user)` in ShopCartService and implement the logic

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -31,6 +34,11 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         user.setPassword(passwordEncoder.encode(user.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleType.USER)));
-        return userMapper.toDto(userRepository.save(user));
+        user = userRepository.save(user);
```
```suggestion
        userRepository.save(user);
```

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Fixed some mistakes, Changed CartItemDto response to ShoppingCartDto in add and update methods, also added Transactional annotation to this methods. Changed add method in CartServiceImpl.

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -31,6 +33,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         user.setPassword(passwordEncoder.encode(user.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleType.USER)));
-        return userMapper.toDto(userRepository.save(user));
+        cartService.createShoppingCart(userRepository.save(user));
```
```suggestion
        userRepository.save(user)
        cartService.createShoppingCart(user);
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    @Transactional
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+        shoppingCart.setUser(userRepository.findById(user.getId()).orElseThrow(() ->
+                new EntityNotFoundException("Not found user with id: " + user.getId())
+        ));
+        cartRepository.save(shoppingCart);
```
```suggestion
        shoppingCart.setUser(user);
        cartRepository.save(shoppingCart);
```

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Removed Transactional from createShoppingCart method.
Split set user and save shoppingCart on 2 different events. And fix 1 more mistake.

okuzan commented on src/main/java/org/example/bookshop/dto/CartItemDto.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
```
Not a mistake but FYI it's ok to use @Data from lombok on DTOs, just not JPA entities

okuzan commented on src/main/java/org/example/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public void delete(Long id, User user) {
+        if (cartItemRepository.findByIdAndUserId(id, user.getId()).isPresent()) {
+            cartItemRepository.deleteById(id);
+        } else {
+            throw new EntityNotFoundException("Can't find cartItem by id: " + id);
+        }
```
```suggestion
        if (cartItemRepository.findByIdAndUserId(id, user.getId()).isEmpty()) {
            throw new EntityNotFoundException("Can't find cartItem by id: " + id);
        }
        cartItemRepository.deleteById(id);
```

okuzan replied:
we don't need else after returning throwing, so in such cases, to simplify the code, invert if condition

dmarmul committed:
Improved solution
